src/support/mtx_rw.c

```diff
@@ -214,7 +214,7 @@ __wt_readlock(WT_SESSION_IMPL *session, WT_RWLOCK *l)
          * until we can get a valid ticket.
          */
         writers_active = old.u.s.next - old.u.s.current;
-        if (old.u.s.readers_queued > writers_active) {
```

Comment:
I think this is correct, but this code is subtle and we should look out for performance fallout from this change.

Like the comment says, it's trying to limit how big the next group of readers can be.  Those readers will all be allowed into the lock together.  The heuristic we're using here is that if there are `N` writers in line to enter the lock, only allow the group of readers after them to grow to `N`.

The bug was that there is also a limit of 255 readers because the counter is a `uint8_t` and this check was also supposed to prevent overflow.

Your change means that when there are `N` writers, we will only allow the next group of readers to be `N - 1`.  What I'm worried about is the behaviour when there is exactly one writer -- all read requests will spin (and another writer could potentially queue itself) where previously we would have queued one of the readers.

That's why I suggested the more complex check `old.u.s.readers_queued > writers_active || old.u.s.readers_queued == 255` -- it more directly avoids overflow and shouldn't have any surprising performance impacts.

Reply:
Another way to do this would be down at line 228 to have something like this:
```
if (++new.u.s.readers_queued == 0)
    goto stall;
else if (new.u.s.readers_queued == 1)
    new.u.s.reader = new.u.s.next;
...
```

Reply:
@michaelcahill: I did actually spent some time understanding the potential fallout of reducing the number of readers by one. When I read the comment in the code `don't allow more readers to queue than there are active writers (calculated as next - current)`, I interpreted it as readers waiting for lock should not exceed total number of writers _holding_ the lock and _waiting_ for lock. For example, taking your example of one writer, i.e., `next = 1` and `curr = 0`, existing code will allow `readers_queued` value to reach 2 before blocking the incoming readers. This means that we will have one active writer and two waiting readers. To generalise, this means for every `N` writers(active + waiting), there are `N+1` readers waiting. With my change, you will only have one active writer and one waiting reader, essentially meaning there are `N` readers for `N` writers(active + waiting). 

One advantage of your proposed change is that we will match the existing heuristic so we should avoid any subtle changes in performance (for better or worse).

Reply:
@hbokhari, thanks, got it.  I still think it would be preferable to make the overflow check explicit, and only change the queuing heuristic after measuring the impact.

Reply:
@michaelcahill : I agree that we shouldn't change the heuristic without measuring the impact. I will submit changes with just an additional check for overflow.